### ISSUE TITLE
Segmentation fault when passing byref parameters to tailcall methods

### ISSUE THREAD

**Issue by eiriktsarpalis** [dotnet/runtime#58531]
### Description

I came across a segmentation fault when working on some F# code. A minimal reproduction showcasing the bug can be found [in this repo](https://github.com/eiriktsarpalis/SegmentationFaultRepro/tree/master). The following error is thrown when trying to run the console app using release builds:
```
Fatal error. Internal CLR error. (0x80131506)
   at System.Array.GetEnumerator()
   at Program.<Main>$(System.String[])
Segmentation fault
```
I believe the issue is triggered by the F# compiler inserting a `tail` opcode in the [`Get` method](https://github.com/eiriktsarpalis/SegmentationFaultRepro/blob/cbf562954b969c44572086a1fe988b7a5dc4a768/ClassLibrary/Getter.fs#L7). Relevant IL:
```
.method public hidebysig 
	instance !FieldType Get (
		!DeclaringType 'instance'
	) cil managed 
{
	// Method begins at RVA 0x2064
	// Code size 16 (0x10)
	.maxstack 8

	// return getter(in instance);
	IL_0000: ldarg.0
	IL_0001: ldfld class ClassLibrary.Getter`2<!0, !1> class ClassLibrary.GetterWrapper`2<!DeclaringType, !FieldType>::getter
	IL_0006: ldarga.s 'instance'
	IL_0008: tail.
	IL_000a: callvirt instance !1 class ClassLibrary.Getter`2<!DeclaringType, !FieldType>::Invoke(!0& modreq([System.Runtime]System.Runtime.InteropServices.InAttribute))
	IL_000f: ret
} // end of method GetterWrapper`2::Get
```

### Configuration

* Which version of .NET is the code running on? _6.0.100-rc.2.21426.20_
* What OS and version, and what distro if applicable? Windows 10 21H1. Does not reproduce on Linux x64.
* What is the architecture? x64. Windows arm64 surfaces a nullref exception instead.
* Do you know whether it is specific to that configuration?
* If you're using Blazor, which web browser(s) do you see this issue in?

### Regression?

Not a regression, this reproduces in .NET 5 as well.

**Comment by jakobbotsch** [comment]
This looks like illegal IL. Passing the address of a local variable to a tailcall is not legal. cc @dotnet/fsharp 
The JIT takes `tail.` prefix as a hint that the IL producer has done the necessary analysis and disables hazard checks, in this case the fact that the parameter is address exposed.

**Comment by dsyme** [comment]
Yes that shouldn't be happening, though we have extensive checks for this and I'm surprised they aren't suppressing the `tail.`

I will transfer to dotnet/fsharp
